odl/discr/discr_utils.py

```diff
@@ -0,0 +1,812 @@
+
+
+class _NearestInterpolator(_Interpolator):
+
+    """Nearest neighbor interpolator.
+
+    The code is adapted from SciPy's `RegularGridInterpolator
+    <http://docs.scipy.org/doc/scipy/reference/generated/\
+scipy.interpolate.RegularGridInterpolator.html>`_ class.
+
+    This implementation is faster than the more generic one in the
+    `_PerAxisPointwiseInterpolator`. Compared to the original code,
+    support of ``'left'`` and ``'right'`` variants are added.
+    """
+
+    def __init__(self, coord_vecs, values, input_type, variant):
+        """Initialize a new instance.
+
+        coord_vecs : sequence of `numpy.ndarray`'s
+            Coordinate vectors defining the interpolation grid
+        values : `array-like`
+            Grid values to use for interpolation
+        input_type : {'array', 'meshgrid'}
+            Type of expected input values in ``__call__``
+        variant : {'left', 'right'}
+            Indicates which neighbor to prefer in the interpolation
+        """
+        super(_NearestInterpolator, self).__init__(
+            coord_vecs, values, input_type)
+        variant_ = str(variant).lower()
+        if variant_ not in ('left', 'right'):
+            raise ValueError("variant '{}' not understood".format(variant_))
+        self.variant = variant_
+
+    def _evaluate(self, indices, norm_distances, out=None):
+        """Evaluate nearest interpolation."""
+        idx_res = []
+        for i, yi in zip(indices, norm_distances):
+            if self.variant == 'left':
+                idx_res.append(np.where(yi <= .5, i, i + 1))
+            else:
+                idx_res.append(np.where(yi < .5, i, i + 1))
+        idx_res = tuple(idx_res)
+        if out is not None:
+            out[:] = self.values[idx_res]
+            return out
+        else:
+            return self.values[idx_res]
+
+
+def _compute_nearest_weights_edge(idcs, ndist, variant):
+    """Helper for nearest interpolation mimicing the linear case."""
+    # Get out-of-bounds indices from the norm_distances. Negative
+    # means "too low", larger than or equal to 1 means "too high"
+    lo = (ndist < 0)
+    hi = (ndist > 1)
+
+    # For "too low" nodes, the lower neighbor gets weight zero;
+    # "too high" gets 1.
+    if variant == 'left':
+        w_lo = np.where(ndist <= 0.5, 1.0, 0.0)
+    else:
+        w_lo = np.where(ndist < 0.5, 1.0, 0.0)
+
+    w_lo[lo] = 0
+    w_lo[hi] = 1
+
+    # For "too high" nodes, the upper neighbor gets weight zero;
+    # "too low" gets 1.
+    if variant == 'left':
+        w_hi = np.where(ndist <= 0.5, 0.0, 1.0)
+    else:
+        w_hi = np.where(ndist < 0.5, 0.0, 1.0)
+
+    w_hi[lo] = 1
+    w_hi[hi] = 0
+
+    # For upper/lower out-of-bounds nodes, we need to set the
+    # lower/upper neighbors to the last/first grid point
+    edge = [idcs, idcs + 1]
+    edge[0][hi] = -1
+    edge[1][lo] = 0
+
+    return w_lo, w_hi, edge
+
+
+def _compute_linear_weights_edge(idcs, ndist):
```

Comment:
Is this where the implicit assumption is made on what happens to the value at interpolation points that are, e.g., larger than the current largest grid point?

Reply:
Yes, I hoped that the comments explain it in a somewhat understandable way. What are you after specifically?

Reply:
This was related to my question above about how the linear interpolation is done for points outside of the actual grid. However, I fail to see the difference between `weight zero` and `value zero` - don't they both achieve the same result about an implicit boundary value of `0`?

(At least in 1D, but maybe there is something in higher dimensions I do not fully grasp)

Reply:
You have the right hunch: in nD, the weights are the tensor product of the 1D weights.
So for instance, in 2D, if a weight is zero in one dimension, the total weights will have a zero row or column due to that. On the other hand, setting a single value to 0 has a very different effect.